### ISSUE TITLE
Replace UTF code

### ISSUE THREAD

**Issue by rsmmr** [zeek/zeek#1756]
The library code that we currently use for UTF8 processing (in `src/Convert.{h,cc}` is cumbersome to use, and lacks some higher-level functionality too (e.g., iterating through a sequence of UTF8 code points). We should replace this with something else, and I think a good candidate is [utf8proc](https://juliastrings.github.io/utf8proc/), which is small, dependency-free library with a simple API. Spicy uses that already, and it works very well.

**Comment by matthewluckie** [comment]
I have looked into this one.  I am not hugely familiar with utf8proc so am happy to have feedback.  I haven't touched the code in binpac_zeek-lib.pac yet.

The first patch works and passes zeek --test (patch and test -utf8proc-1.txt).  The second patch uses utf8proc_category() and associated enum instead of the magic hexadecimal numbers, but doesn't pass tests (patch and test -utf8proc-2.txt)

Should I be aiming to use utf8proc_category()?

[patch-utf8proc-1.txt](https://github.com/zeek/zeek/files/8183756/patch-utf8proc-1.txt)
[patch-utf8proc-2.txt](https://github.com/zeek/zeek/files/8183757/patch-utf8proc-2.txt)
[test-utf8proc-1.txt](https://github.com/zeek/zeek/files/8183758/test-utf8proc-1.txt)
[test-utf8proc-2.txt](https://github.com/zeek/zeek/files/8183760/test-utf8proc-2.txt)

**Comment by timwoj** [comment]
This might be worth looking at in conjunction with this https://github.com/uni-algo/uni-algo

**Comment by varshneydevansh** [comment]
Hi @timwoj I am new to Zeek project and looking to start contributing to it. As back in 2017, Computer Networks did catch my attention and eventually ended up making my very first file sharing and chat cross-platform app with Qt.

So, can I start working from this challenge or #2897 ?

**Comment by timwoj** [comment]
@varshneydevansh Certainly, I've assigned this to you. I hadn't had a chance to look at it anyways.

**Comment by varshneydevansh** [comment]
Will work on this soon =)

**Comment by ckreibich** [comment]
@varshneydevansh are you still planning to look at this? If not, perhaps @matthewluckie is still up for it after all this time. :-)

**Comment by varshneydevansh** [comment]
Hi @ckreibich you can assign him as I am currently working on my book. I will pick another soon, completing my writing task. =)

**Comment by ckreibich** [comment]
Thanks for the quick response. Let me unassign you and we'll see if Matthew is still around.

**Comment by matthewluckie** [comment]
Hi, I don't have cycles for this at this point in time.  I submitted two patches when utf8proc was the suggested approach, and asked for feedback.  the second patch uses utf8proc enums that define the various categories that the spec uses, but these seemed to clash with the current hard-coded ranges.
